### PR TITLE
OP-26699 Hyperclair replacement - WIP

### PR DESCRIPTION
**Pull request by murat1985**
So far just brain dump to fix mistakes at initial stage. Have tested methods separately seems working. 

TODO: 
* implement reporting 
* clean out code and check edge cases
* tests

/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team

### References
 - Jira link: https://zendesk.atlassian.net/browse/OP-26699

### Risks
- Level: Low

@grosser

### CODE REVIEW

File: plugins/clair/clair.gemspec [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+Gem::Specification.new "samson_hyperclair", "0.0.0" do |s|
+  s.summary = "Samson Hyperclair integration"
+  s.authors = ["Michael Grosser"]
+  s.email = "mgrosser@zendesk.com"
```
Comment (grosser) [inline plugins/clair/clair.gemspec]: don't blame me for this :D

File: plugins/clair/lib/samson_clair/samson_plugin.rb [hunk 2]
```diff
@@ -0,0 +1,235 @@
+# frozen_string_literal: true
+# FIXME: check based on docker_repo_digest not tag
```
Comment (grosser) [inline plugins/clair/lib/samson_clair/samson_plugin.rb]: remove that ?

**Comment by grosser** [comment]
this whole clair logic smells like it should be a gem since it's not really samson specific

File: plugins/clair/lib/samson_clair/samson_plugin.rb [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        job.reload
+        job.update_column(:output, job.output + output)
+      else
+        Rails.logger.info(output)
+      end
+    end
+
+    def parse_digest(docker_repo_digest)
+      regexp = /^(?:([^\/]+)\/)?(?:(.+)\/)*([^@:\/]+)(?:[@:](.+))?/
+      docker_repo_digest.scan(regexp).first
```
Comment (grosser) [inline plugins/clair/lib/samson_clair/samson_plugin.rb]: same as `docker_repo_digest[/^(?:([^\/]+)\/)?(?:(.+)\/)*([^@:\/]+)(?:[@:](.+))?/]`

File: plugins/clair/lib/samson_clair/samson_plugin.rb [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    end
+
+    def extract_realm_service(bearer)
+      regexp = /(?:(\w+)=([\w.:\/-]+))/
+      Hash[*bearer.gsub('"','').scan(regexp).flatten]
```
Comment (grosser) [inline plugins/clair/lib/samson_clair/samson_plugin.rb]: use `tr` for single character replacements

File: plugins/clair/lib/samson_clair/samson_plugin.rb [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    end
+
+    def docker_login_pass
+      registry = Docker.first
```
Comment (grosser) [inline plugins/clair/lib/samson_clair/samson_plugin.rb]: `DockerRegistry.first` ?

**Comment by grosser** [comment]
looks not too crazy ... how about you extract all the `clair` stuff into a single class:

`Clair.scan(repo_digest, output: output)` ... that returns true/false or number of vulenerabilities ... then I can grab that and make a gem from it ... and we can have nice integration tests there and only minimal integration code in samson ...

**Comment by murat1985** [comment]
Sounds good, will do

**Comment by jonmoter** [comment]
This PR has been open for over 6 months now. Should we close this out, or is there still a chance work will happen on it?
